File: app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

#[Fillable([
    'client_id', 'vendor_id', 'courier_id', 'status',
    'delivery_latitude', 'delivery_longitude', 'delivery_address_text',
    'total_amount', 'delivery_fee', 'commission_amount', 'source', 'promo_code_used',
])]
class Order extends Model
{
    public const STATUS_LABELS = [
        'en_attente_paiement' => 'En attente de paiement',
        'paiement_sequestre' => 'Paiement séquestré',
        'confirmee' => 'Confirmée',
        'en_preparation' => 'En préparation',
        'cherche_livreur' => 'Recherche livreur',
        'livreur_assigne' => 'Livreur assigné',
        'recuperee' => 'Récupérée',
        'en_livraison' => 'En livraison',
        'livree' => 'Livrée',
        'paiement_libere' => 'Paiement libéré',
        'litige_ouvert' => 'Litige ouvert',
        'annulee' => 'Annulée',
    ];

    protected static function booted(): void
    {
        // The DB column defaults to 'en_attente_paiement', but that default
        // isn't reflected on the in-memory model right after insert — set it
        // explicitly so OrderObserver sees the real value when logging history.
        static::creating(function (Order $order) {
            $order->status ??= 'en_attente_paiement';
        });
    }

    protected function casts(): array
    {
        return [
            'delivery_latitude' => 'decimal:7',
            'delivery_longitude' => 'decimal:7',
            'total_amount' => 'decimal:2',
            'delivery_fee' => 'decimal:2',
            'commission_amount' => 'decimal:2',
        ];
    }

    public function client(): BelongsTo
    {
        return $this->belongsTo(User::class, 'client_id');
    }

    public function vendor(): BelongsTo
    {
        return $this->belongsTo(Vendor::class);
    }

    public function courier(): BelongsTo
    {
        return $this->belongsTo(User::class, 'courier_id');
    }

    public function items(): HasMany
    {
        return $this->hasMany(OrderItem::class);
    }

    public function statusHistory(): HasMany
    {
        return $this->hasMany(OrderStatusHistory::class);
    }

    public function payments(): HasMany
    {
        return $this->hasMany(Payment::class);
    }
}

File: mobile/lib/models/order.dart
class OrderItem {
  final int id;
  final int listingId;
  final String? listingName;
  final int quantity;
  final double unitPrice;

  OrderItem({
    required this.id,
    required this.listingId,
    required this.quantity,
    required this.unitPrice,
    this.listingName,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      id: json['id'],
      listingId: json['listing_id'],
      listingName: json['listing_name'],
      quantity: json['quantity'],
      unitPrice: double.parse(json['unit_price'].toString()),
    );
  }
}

/// Mirrors Order::STATUS_LABELS on the backend.
const Map<String, String> orderStatusLabels = {
  'en_attente_paiement': 'En attente de paiement',
  'paiement_sequestre': 'Paiement séquestré',
  'confirmee': 'Confirmée',
  'en_preparation': 'En préparation',
  'cherche_livreur': 'Recherche livreur',
  'livreur_assigne': 'Livreur assigné',
  'recuperee': 'Récupérée',
  'en_livraison': 'En livraison',
  'livree': 'Livrée',
  'paiement_libere': 'Paiement libéré',
  'litige_ouvert': 'Litige ouvert',
  'annulee': 'Annulée',
};

class Order {
  final int id;
  final int clientId;
  final int vendorId;
  final int? courierId;
  final String status;
  final String deliveryAddressText;
  final double totalAmount;
  final double deliveryFee;
  final String source;
  final String? vendorBusinessName;
  final String? vendorAddressText;
  final String? clientName;
  final DateTime? createdAt;
  final List<OrderItem> items;

  Order({
    required this.id,
    required this.clientId,
    required this.vendorId,
    required this.status,
    required this.deliveryAddressText,
    required this.totalAmount,
    required this.deliveryFee,
    required this.source,
    this.courierId,
    this.vendorBusinessName,
    this.vendorAddressText,
    this.clientName,
    this.createdAt,
    this.items = const [],
  });

  String get statusLabel => orderStatusLabels[status] ?? status;

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'],
      clientId: json['client_id'],
      vendorId: json['vendor_id'],
      courierId: json['courier_id'],
      status: json['status'],
      deliveryAddressText: json['delivery_address_text'] ?? '',
      totalAmount: double.parse(json['total_amount'].toString()),
      deliveryFee: double.parse((json['delivery_fee'] ?? 0).toString()),
      source: json['source'] ?? 'app',
      vendorBusinessName: json['vendor_business_name'],
      vendorAddressText: json['vendor_address_text'],
      clientName: json['client_name'],
      createdAt: json['created_at'] != null ? DateTime.tryParse(json['created_at']) : null,
      items: (json['items'] as List<dynamic>? ?? [])
          .map((item) => OrderItem.fromJson(item))
          .toList(),
    );
  }
}
